game/views.py

```diff
@@ -36,3 +37,24 @@ class StartGameSessionView(APIView):
     def post(self, request, *args, **kwargs):
         GameSessionService.start_game_session(request.user)
         return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+class UpgradeBuildingView(APIView):
+    def post(self, request, name, *args, **kwargs):
+        village = request.user.village
+        building = village.get_building(name)
+        upgrade_costs = building.get_upgrade_cost()
+        village_resources = village.resources
+
+        for resource in upgrade_costs:
+            if village_resources[resource] < upgrade_costs[resource]:
+                raise exceptions.ResourcesException
+
+        for resource in upgrade_costs:
+            village_resources[resource] -= upgrade_costs[resource]
+
+        building.upgrade()
+        village.upgrade_building_level(name)
+        village.save()
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
```

Comment:
If you want you can extract the logic to separate service `VillageService.upgrade_building(player, building_name)`